### PR TITLE
Add `.setPriority()` for updating priority of a queued promise function

### LINKED ISSUE #208: Is it feasible to change priority of a Queued item on the fly
I'm working on a custom loader with asynchronous load promises, I want to change the priority of a specific promise function, can this be achieved using current code? 

I managed to achieve something by queuing up the promise function again with updated priority, but this loads the item twice. 

```ts
prioritize(key: string) {
    const promise = this.taskList.get(key);
    if (!promise) {
      console.error("No such promise: ", key);
    } else {
      this.pause();
      this.queue.add(promise, { priority: 2 });
      this.resume();
    }
}
```
Thanks.

### PR DESCRIPTION
**Pull request by RaishavHanspal**
Fixes #208 - add a `uid` to track the promise functions and update the priority on any promise.
also add a new event `started` to notify when a promise function is executed.

### CODE REVIEW

File: source/index.ts [hunk 1]
```diff
@@ -8,7 +8,7 @@ type Task<TaskResultType> =
 	| ((options: TaskOptions) => PromiseLike<TaskResultType>)
 	| ((options: TaskOptions) => TaskResultType);
 
-type EventName = 'active' | 'idle' | 'empty' | 'add' | 'next' | 'completed' | 'error';
+type EventName = 'active' | 'idle' | 'empty' | 'add' | 'next' | 'completed' | 'error' | 'started';
```
Comment (sindresorhus) [inline source/index.ts:11]: Don't make unrelated changes. This should be a separate pull request after this one.

File: source/index.ts [hunk 2]
```diff
@@ -228,15 +231,25 @@ export default class PQueue<QueueType extends Queue<RunFunction, EnqueueOptionsT
 		});
 	}
 
+	async setPriority(uid: string, priority: number) {
```
Comment (sindresorhus) [inline source/index.ts:234]: I think we can just call it `id`.

Comment (sindresorhus) [reply source/index.ts:234]: Why is this async?

File: source/index.ts [hunk 3]
```diff
@@ -228,15 +231,25 @@ export default class PQueue<QueueType extends Queue<RunFunction, EnqueueOptionsT
...
+
 	/**
 	Adds a sync or async task to the queue. Always returns a promise.
 	*/
-	async add<TaskResultType>(function_: Task<TaskResultType>, options: {throwOnTimeout: true} & Exclude<EnqueueOptionsType, 'throwOnTimeout'>): Promise<TaskResultType>;
-	async add<TaskResultType>(function_: Task<TaskResultType>, options?: Partial<EnqueueOptionsType>): Promise<TaskResultType | void>;
-	async add<TaskResultType>(function_: Task<TaskResultType>, options: Partial<EnqueueOptionsType> = {}): Promise<TaskResultType | void> {
+	async add<TaskResultType>(function_: Task<TaskResultType>, options: {throwOnTimeout: true} & Exclude<EnqueueOptionsType, 'throwOnTimeout'>, uid?: string): Promise<TaskResultType>;
+	async add<TaskResultType>(function_: Task<TaskResultType>, options?: Partial<EnqueueOptionsType>, uid?: string): Promise<TaskResultType | void>;
+	async add<TaskResultType>(function_: Task<TaskResultType>, options: Partial<EnqueueOptionsType> = {}, uid?: string): Promise<TaskResultType | void> {
```
Comment (sindresorhus) [inline source/index.ts:243]: What's the use-case for a custom ID? I would prefer not to expose this unless there are real-world use-cases.

File: source/priority-queue.ts [hunk 4]
```diff
@@ -32,6 +33,27 @@ export default class PriorityQueue implements Queue<RunFunction, PriorityQueueOp
 		this.#queue.splice(index, 0, element);
 	}
 
+	prioritize(uid: string, priority?: number) {
```
Comment (sindresorhus) [inline source/priority-queue.ts:36]:
```suggestion
	setPriority(uid: string, priority?: number) {
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: source/index.ts (hunk 1 above)
Comment (RaishavHanspal) [reply source/index.ts]: reverted from current PR

File: source/index.ts (hunk 2 above)
Comment (RaishavHanspal) [reply source/index.ts]: removed unnecessary async

File: source/index.ts (hunk 3 above)
Comment (RaishavHanspal) [reply source/index.ts]: shifted the id in `EnqueueOptionsType`. use case is simple a unique identifier that can be assigned to query in the queue.

File: source/priority-queue.ts (hunk 4 above)
Comment (RaishavHanspal) [reply source/priority-queue.ts]: updated the suggestion

File: source/index.ts (hunk 3 above)
Comment (sindresorhus) [reply source/index.ts]: The problem with both accepting an ID and generating is that the user could potentially pass in an ID that equals one of the existing IDs generated internally.

Comment (RaishavHanspal) [reply source/index.ts]: we could prevent user to change/assign the id and completely rely on the auto generated id - by idAssigner, but we would need to devise a way to emit this info as soon as the id is generated? could you please suggest on this?

File: source/priority-queue.ts [hunk 5]
```diff
@@ -32,6 +32,27 @@ export default class PriorityQueue implements Queue<RunFunction, PriorityQueueOp
 		this.#queue.splice(index, 0, element);
 	}
 
+	setPriority(id: string, priority?: number) {
+		const existingIndex: number = this.#queue.findIndex((element: Readonly<PriorityQueueOptions>) => element.id === id);
+		const [item] = this.#queue.splice(existingIndex, 1);
```
Comment (Richienb) [inline source/priority-queue.ts:37]: Please add tests for when an invalid id is passed.

I think this code has a bug where the `-1` returned by `Array#findIndex` causes the last item to be spliced, instead of no item at all.

File: source/priority-queue.ts [hunk 6]
```diff
@@ -32,6 +32,27 @@ export default class PriorityQueue implements Queue<RunFunction, PriorityQueueOp
...
+		if (item === undefined) {
+			return;
+		}
+
+		item.priority = priority ?? ((item.priority ?? 0) + 1);
```
Comment (Richienb) [inline source/priority-queue.ts:42]: In my opinion, `.setPriority('abc')` does not make it obvious enough that it increases the priority by `1`. What about if one wanted to decrease priority. Maybe more methods might be needed?

File: source/priority-queue.ts [hunk 7]
```diff
@@ -32,6 +32,27 @@ export default class PriorityQueue implements Queue<RunFunction, PriorityQueueOp
...
+		if (this.size && this.#queue[this.size - 1]!.priority! >= priority!) {
```
Comment (Richienb) [inline source/priority-queue.ts:43]: It looks like the check for `this.size` ensures that it exists. However, it's already a required property:

https://github.com/sindresorhus/p-queue/blob/47a4f19ab3d4422e06cb1b618f4c45a11b5def30/source/queue.ts#L4

So, if the check fails, we have bigger problems and might even want to throw an error.

This shortcut would also be incorrectly passed by if `this.size === 0`.

```suggestion
		if (this.size === 0 || this.#queue[this.size - 1]!.priority! >= priority!) {
```

**Review by Richienb — CHANGES_REQUESTED** [review]

File: source/priority-queue.ts [hunk 8]
```diff
@@ -32,6 +32,27 @@ export default class PriorityQueue implements Queue<RunFunction, PriorityQueueOp
 		this.#queue.splice(index, 0, element);
 	}
 
+	setPriority(id: string, priority?: number) {
+		const existingIndex: number = this.#queue.findIndex((element: Readonly<PriorityQueueOptions>) => element.id === id);
+		const [item] = this.#queue.splice(existingIndex, 1);
+		if (item === undefined) {
+			return;
+		}
```
Comment (Richienb) [inline source/priority-queue.ts:40]: Might fix bug
```suggestion
		if (existingIndex === -1) {
			return;
		}
		const [item] = this.#queue.splice(existingIndex, 1);
```

File: source/priority-queue.ts (hunk 7 above)
Comment (RaishavHanspal) [reply source/priority-queue.ts:43]: sounds right! shall we update this at #L23 too, since this change is taken from there, or should we create a common function to position the promise function at right place in queue wrt priority?
Please do suggest a better name for this function in case we want to move with latter. :)

File: source/priority-queue.ts (hunk 6 above)
Comment (RaishavHanspal) [reply source/priority-queue.ts:42]: incrementing the priority by 1, might not be a generic use-case, so reverting it to be user specified.

File: source/priority-queue.ts (hunk 7 above)
Comment (Richienb) [reply source/priority-queue.ts]: > should we create a common function to position the promise function at right place in queue wrt priority?
Let create a private `_enqueueElement` method.

File: source/priority-queue.ts [hunk 9]
```diff
@@ -32,6 +33,31 @@ export default class PriorityQueue implements Queue<RunFunction, PriorityQueueOp
...
 
+	setPriority(id: string, priority: number) {
+		const existingIndex: number = this.#queue.findIndex((element: Readonly<PriorityQueueOptions>) => element.id === id);
+		if (existingIndex === -1) {
+			throw new Error('Invalid Index - No promise function of specified id available in the queue.');
+		}
+
+		const [item] = this.#queue.splice(existingIndex, 1);
+		if (item === undefined) {
+			return;
```
Comment (Richienb) [inline source/priority-queue.ts:44]: Is this state reachable?

**Review by Richienb — CHANGES_REQUESTED** [review]

File: source/priority-queue.ts (hunk 9 above)
Comment (RaishavHanspal) [reply source/priority-queue.ts]: As per the existing linting settings, compiler suggests an error - of possible undefined value. Added a throw Error statement for the same.

File: source/priority-queue.ts (hunk 7 above)
Comment (RaishavHanspal) [reply source/priority-queue.ts]: reused `enqueue` method itself as there was no additional code in the same.

**Comment by RaishavHanspal** [comment]
requested changes have been pushed! @Richienb

**Comment by Richienb** [comment]
I think we should allow searching by the object returned by enqueue as well. What do you think? Does there remain a use case where an id string remains more useful?

**Comment by RaishavHanspal** [comment]
@Richienb that would be a very nice idea, and will not require `id` property. The problem that I faced was enqueue accepts `RunFunction` - which in my understanding is a callback not accessible to the end-user.
The Idea of priority update - is for the user to be able to update the priority on the fly, currently this is achieved by a unique `id` assigned by user, or this can be passed using the `invoke` event as added in #210 incase assigned by `#idAssigner`.

**Comment by RaishavHanspal** [comment]
Hi @Richienb, @sindresorhus , Please suggest further.

File: test/test.ts [hunk 10]
```diff
@@ -1134,3 +1134,43 @@ test('aborting multiple jobs at the same time', async t => {
...
+	queue.add(async () => {
+		await delay(400);
+		result.push('🐢');
+	}, {id: '🐢'});
+	queue.setPriority('🐢', 1);
+	await queue.onIdle();
+	t.deepEqual(result, ['🐌', '🐢', '🦆']);
+});
+
+test.failing('.setPriority() - with invalid "id"', async t => {
```
Comment (sindresorhus) [inline test/test.ts:1158]: `test.failing` is not for tests that are intentionally failing.

File: source/priority-queue.ts [hunk 11]
```diff
@@ -32,6 +33,20 @@ export default class PriorityQueue implements Queue<RunFunction, PriorityQueueOp
...
+	setPriority(id: string, priority: number) {
+		const existingIndex: number = this.#queue.findIndex((element: Readonly<PriorityQueueOptions>) => element.id === id);
+		if (existingIndex === -1) {
+			throw new Error('Invalid Index - No promise function of specified id available in the queue.');
+		}
+
+		const [item] = this.#queue.splice(existingIndex, 1);
+		if (item === undefined) {
+			throw new Error('Undefined Item - No promise function of specified id available in the queue.');
+		}
```
Comment (sindresorhus) [inline source/priority-queue.ts:45]: I don't think these should throw. It should be allowed to set a priority without knowing beforehand whether the has already finished.

File: source/options.ts [hunk 12]
```diff
@@ -69,6 +69,7 @@ export type QueueAddOptions = {
 	@default 0
 	*/
 	readonly priority?: number;
+	id?: string;
```
Comment (sindresorhus) [inline source/options.ts:72]: Doc comment.

File: source/index.ts [hunk 13]
```diff
@@ -228,12 +231,21 @@ export default class PQueue<QueueType extends Queue<RunFunction, EnqueueOptionsT
...
+		this.#queue.setPriority(id, priority);
+	}
+
 	/**
 	Adds a sync or async task to the queue. Always returns a promise.
 	*/
 	async add<TaskResultType>(function_: Task<TaskResultType>, options: {throwOnTimeout: true} & Exclude<EnqueueOptionsType, 'throwOnTimeout'>): Promise<TaskResultType>;
 	async add<TaskResultType>(function_: Task<TaskResultType>, options?: Partial<EnqueueOptionsType>): Promise<TaskResultType | void>;
 	async add<TaskResultType>(function_: Task<TaskResultType>, options: Partial<EnqueueOptionsType> = {}): Promise<TaskResultType | void> {
+		// Incase id is not defined
```
Comment (sindresorhus) [inline source/index.ts:244]:
```suggestion
		// In case `id` is not defined.
```

File: source/index.ts [hunk 14]
```diff
@@ -228,12 +231,21 @@ export default class PQueue<QueueType extends Queue<RunFunction, EnqueueOptionsT
 		});
 	}
 
+	setPriority(id: string, priority: number) {
```
Comment (sindresorhus) [inline source/index.ts:234]: Doc comment

File: source/index.ts [hunk 15]
```diff
@@ -43,6 +43,9 @@ export default class PQueue<QueueType extends Queue<RunFunction, EnqueueOptionsT
 
 	readonly #throwOnTimeout: boolean;
 
+	/** Use to assign a unique identifier to a promise function, if not explicitly specified */
```
Comment (sindresorhus) [inline source/index.ts:46]:
```suggestion
	// Used to assign a unique identifier to a promise function, if not explicitly specified.
```

File: source/index.ts [hunk 16]
```diff
@@ -43,6 +43,9 @@ export default class PQueue<QueueType extends Queue<RunFunction, EnqueueOptionsT
...
+	#idAssigner = 1;
```
Comment (sindresorhus) [inline source/index.ts:47]: I think this should be a BigInt, to ensure it will never overflow.

File: source/index.ts [hunk 17]
```diff
@@ -228,12 +231,21 @@ export default class PQueue<QueueType extends Queue<RunFunction, EnqueueOptionsT
...
 	/**
 	Adds a sync or async task to the queue. Always returns a promise.
 	*/
 	async add<TaskResultType>(function_: Task<TaskResultType>, options: {throwOnTimeout: true} & Exclude<EnqueueOptionsType, 'throwOnTimeout'>): Promise<TaskResultType>;
 	async add<TaskResultType>(function_: Task<TaskResultType>, options?: Partial<EnqueueOptionsType>): Promise<TaskResultType | void>;
 	async add<TaskResultType>(function_: Task<TaskResultType>, options: Partial<EnqueueOptionsType> = {}): Promise<TaskResultType | void> {
+		// Incase id is not defined
+		if (options.id === undefined) {
+			options.id = (this.#idAssigner++).toString();
+		}
```
Comment (sindresorhus) [inline source/index.ts:247]:
```suggestion
		options.id ??= (this.#idAssigner++).toString();
```

File: source/options.ts (hunk 12 above)
Comment (sindresorhus) [inline source/options.ts:72]:
```suggestion
	readonly id?: string;
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: source/options.ts (hunk 12 above)
Comment (RaishavHanspal) [reply source/options.ts:72]: This value is getting updated as per:
https://github.com/RaishavHanspal/p-queue/blob/686f0d8e21efc1cae0b4c855d6a1eab58258d436/source/index.ts#L245

File: source/options.ts [hunk 18]
```diff
@@ -69,6 +69,13 @@ export type QueueAddOptions = {
 	@default 0
 	*/
 	readonly priority?: number;
+	/**
+	Unique identifier for the promise function. This can be used to update priority, before it gets executed.
+
+	@default '1'
```
Comment (sindresorhus) [inline source/options.ts:76]: This is not the default value always.

File: source/options.ts [hunk 19]
```diff
@@ -69,6 +69,13 @@ export type QueueAddOptions = {
...
+
+	Value of `id` will be assigned using a bigint number assigner which increments for every new promise function with unspecified id
```
Comment (sindresorhus) [inline source/options.ts:78]: This needs to be better written to make it clear it only does this when no ID is specified.

File: source/index.ts [hunk 20]
```diff
@@ -231,6 +231,9 @@ export default class PQueue<QueueType extends Queue<RunFunction, EnqueueOptionsT
 		});
 	}
 
+	/**
+	Update priority of a known promise function, using the `id` identifier, and a priority value to override existing priority value. The updated value of priority ensures whether to execute this promise function sooner or later.
```
Comment (sindresorhus) [inline source/index.ts:235]: Readme and index.d.ts should be in sync.

File: source/options.ts [hunk 21]
```diff
@@ -69,6 +69,13 @@ export type QueueAddOptions = {
...
+	*/
 	id?: string;
```
Comment (sindresorhus) [inline source/options.ts:80]: This is missing from the readme.

File: readme.md [hunk 22]
```diff
@@ -236,6 +236,44 @@ console.log(queue.sizeBy({priority: 0}));
 //=> 1
 ```
 
+#### .setPriority(id, priority)
+
+Update priority of a known promise function, using the `id` identifier, and a priority value to override existing priority value. The updated value of priority ensures whether to execute this promise function sooner or later.
+
+Function works only when we specify a defined concurrency to change any priorities.
```
Comment (sindresorhus) [inline readme.md:243]: Typo.

And this is a bit wordy. Run it through ChatGPT.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
index.d.ts and readme should be in sync, specifically `setPriority`

**Comment by RaishavHanspal** [comment]
Ok @sindresorhus ! would you suggest I add the code snippets as well in the json doc comment just as we have in the readme? Or please suggest otherwise. 🙂

**Comment by sindresorhus** [comment]
Yes

File: readme.md [hunk 23]
```diff
@@ -137,6 +137,12 @@ Default: `0`
 
 Priority of operation. Operations with greater priority will be scheduled first.
 
+##### id
+
+Type `string`
+
+Unique identifier for the promise function, used to update its priority before execution. If not specified, it is auto-assigned as an incrementing bigint starting from 1n.
```
Comment (Richienb) [inline readme.md:144]:
```suggestion
Unique identifier for the promise function, used to update its priority before execution. If not specified, it is auto-assigned an incrementing BigInt starting from `1n`.
```

File: source/options.ts [hunk 24]
```diff
@@ -69,6 +69,11 @@ export type QueueAddOptions = {
 	@default 0
 	*/
 	readonly priority?: number;
+
+	/**
+	Unique identifier for the promise function, used to update its priority before execution. If not specified, it is auto-assigned as an incrementing bigint starting from 1n.
```
Comment (Richienb) [inline source/options.ts:74]:
```suggestion
	Unique identifier for the promise function, used to update its priority before execution. If not specified, it is auto-assigned an incrementing BigInt starting from `1n`.
```

File: source/priority-queue.ts [hunk 25]
```diff
@@ -32,6 +33,16 @@ export default class PriorityQueue implements Queue<RunFunction, PriorityQueueOp
 		this.#queue.splice(index, 0, element);
 	}
 
+	setPriority(id: string, priority: number) {
+		const existingIndex: number = this.#queue.findIndex((element: Readonly<PriorityQueueOptions>) => element.id === id);
+		if (existingIndex === -1) {
+			throw new Error('Invalid Index - No promise function of specified id available in the queue.');
```
Comment (Richienb) [inline source/priority-queue.ts:39]:
```suggestion
			throw new ReferenceError(`No promise function with the id "${id}" exists in the queue.`);
```

File: source/priority-queue.ts [hunk 26]
```diff
@@ -32,6 +33,16 @@ export default class PriorityQueue implements Queue<RunFunction, PriorityQueueOp
 		this.#queue.splice(index, 0, element);
 	}
 
+	setPriority(id: string, priority: number) {
+		const existingIndex: number = this.#queue.findIndex((element: Readonly<PriorityQueueOptions>) => element.id === id);
```
Comment (Richienb) [inline source/priority-queue.ts:37]: I think we can just call it `index`?

**Review by Richienb — CHANGES_REQUESTED** [review]
This is looking pretty good

File: source/priority-queue.ts (hunk 26 above)
Comment (RaishavHanspal) [reply source/priority-queue.ts:37]: That sounds better! updated all instances of `id` to `index`.

Comment (Richienb) [reply source/priority-queue.ts]: Oh wait that's not what I meant. But ok hmm, @sindresorhus do you think `index` turns out to be a better name?

Comment (RaishavHanspal) [reply source/priority-queue.ts]: Oh right! on checking again I feel `index` would be appropriate if that only indicated ordering position i.e. expected to be only number, however it is a unique identifier that could be a string also.
Also - I missed updating `#idAssigner`. 😅

Comment (RaishavHanspal) [reply source/priority-queue.ts]: Hey @Richienb , I've reverted the change - with the intended `index`. Please do let me know otherwise. 🫡

**Review by Richienb — APPROVED** [review]
(Despite https://github.com/sindresorhus/p-queue/pull/209#discussion_r1903123317)

I'm loving it
